src/ini-load.js

```diff
@@ -56,13 +86,83 @@ export function whenContentIniLoad(
     const promises = [];
     resources.forEach((r) => {
       if (!EXCLUDE_INI_LOAD.includes(r.element.tagName)) {
-        promises.push(r.loadedOnce());
+        promises.push(r.element.whenLoaded());
       }
     });
     return Promise.all(promises);
   });
 }
 
+/**
+ * A new way using direct measurement.
+ *
+ * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+ * @param {boolean=} opt_prerenderableOnly signifies if we are in prerender mode.
+ * @return {!Promise}
+ * @visibleForTesting
+ * TODO(#31915): remove visibility
+ */
+export function whenContentIniLoadInOb(elementOrAmpDoc, opt_prerenderableOnly) {
+  const ampdoc = Services.ampdoc(elementOrAmpDoc);
+  return (
+    ampdoc
+      .signals()
+      // First, wait for the `ready-scan` signal. Waiting for each element
+      // individually is too expensive and `ready-scan` will cover most of
+      // the initially parsed elements.
+      .whenSignal(READY_SCAN_SIGNAL)
+      .then(() => {
+        // Filter elements.
+        const resources = Services.resourcesForDoc(ampdoc);
+        const elements = resources
+          .get()
+          .filter((r) => {
+            if (opt_prerenderableOnly && !r.prerenderAllowed()) {
+              return false;
+            }
+            return !EXCLUDE_INI_LOAD.includes(r.element.tagName);
+          })
+          .map((r) => r.element);
+
+        if (elements.length === 0) {
+          return Promise.resolve([]);
+        }
+
+        // Find intersecting elements.
+        return new Promise((resolve) => {
+          const {win} = ampdoc;
+          const io = new win.IntersectionObserver(
```

Comment:
So we're using this as a way to filter our elements down to just the things that are intersecting?

Reply:
Correct. They are already prefiltered. This moves us away from having to call `getBoundingClientRect()` on each of these elements, which is what we do in the classical version.